### DISCUSSION TITLE
inconsistency in partial formatting with Darker

### DISCUSSION THREAD

**Discussion by Eyobkibret15** [akaihola/darker#529] [Q&A] [answered]
Hello,
I've encountered an issue when using Darker to apply partial formatting to a Python file. following the documentation provided here  https://pypi.org/project/darker/ my goal was to selectively format only specific lines of code.
the original Code which i made initial commit on is
`if True: print('hi')`
`print()`
`if False: print('there)`

- when i change only the first print string it works as expected only suggested to format the first line
<img width="450" alt="just" src="https://github.com/akaihola/darker/assets/64076953/b132c17c-87e9-4b17-b956-1341e47ec8eb">

- however, when I remove the empty `print()` from the 2nd line, Darker diff output includes lines that should not be formatted 
<img width="448" alt="2nd" src="https://github.com/akaihola/darker/assets/64076953/a38b402c-4b17-47c7-8105-52f4cf321e9d">

- moreover,  when i add only a new unformatted print statement `if False: print('new line')`  to the original file darker diff still includes line that should not be formatted 
<img width="459" alt="3rd" src="https://github.com/akaihola/darker/assets/64076953/cc8c4688-3cd5-4e84-a7d0-4da49bf9bf5a">

i am seeking assistance to understand why Darker is recommending these additional changes and how i might configure it to respect the boundaries  of partial formatting ?

**Comment by akaihola** [comment, marked as answer]
Hi @Eyobkibret15, thanks for your report!

Since Black doesn't support partial formatting natively, Darker takes a diff between the original and modified versions of the file and tries to track down which lines correspond to each other, and how they map to the reformatting made by Black. It then takes a conservative approach to apply minimal diff chunks which intersect with detected changes. Unfortunately it's very tricky to get this approach to really minimize the changes since we're not digging down to analyzing the Python code.

So unfortunately this is a known limitation which can't be approved using the current approach with reasonable effort.

Let me know if the documentation can be improved to explain this limitation better!

**Reply by Eyobkibret15** [reply]
Thanks @akaihola for the clarification. It would be great if the documentation could briefly mention this limitation and any possible workarounds. Appreciate your help!

**Reply by akaihola** [reply]
@Eyobkibret15 I'm working on improving the documentation in #566. Review and comments are welcome!

**Reply by akaihola** [reply]
@Eyobkibret15 I sent you an invitation to become a contributor on the Darker repository. If you accept that, I can assign you as a reviewer of #566 so you can double check that the documentation I added is understandable.

**Reply by Eyobkibret15** [reply]
Thank you for the invitation, @akaihola. I’ve accepted the contributor role and will review the proposed changes in #566.

**Reply by akaihola** [reply]
Oops @Eyobkibret15 I already merged #566, but feel free to propose further improvements to the text!
